components/TextFields/src/MDCInputTextField.h

```diff
@@ -0,0 +1,120 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <UIKit/UIKit.h>
+
+/**
+ A UITextField subclass that attempts to do the following:
+
+ - Earnestly interpret and actualize the Material guidelines for text fields, which can be found
+ here: https://material.io/design/components/text-fields.html#outlined-text-field
+
+ - Feel intuitive for someone used to the conventions of iOS development and UIKit controls.
+
+ - Enable easy set up and reliable and predictable behavior.
+
+ */
+@interface MDCInputTextField : UITextField
+
+/**
+ The @c floatingLabel is a label that occupies the area the text usually occupies when there is no
+ text and that floats above the text once there is text. It is distinct from the placeholder.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *floatingLabel;
+
+/**
+ When set to YES, the floating label floats when the view becomes the first responder. When
+ set to NO it disappears.
+
+ @note The default is YES.
+ */
+@property(nonatomic, assign) BOOL canFloatingLabelFloat;
```

Comment:
Define this behavior as an enum. It's confusing as a boolean because the label still does something when this is false.

```objc
@property(nonatomic, assign) MDCInputTextFieldLabelBehavior labelBehavior;
```

Reply:
Since other contained input views in disguise also have this behavior what do you think of a type name that could apply to all of them? Maybe like `MDCTextControlLabelBehavior`?

```objc
typedef NS_ENUM(NSInteger, MDCTextControlLabelBehavior) {
  /** Indicates that the text control's label floats when appropriate. */
  MDCTextControlLabelBehaviorFloating,
  /** Indicates that the text control's label does not float. */
  MDCTextControlLabelBehaviorNonFloating,
};
```
or maybe:
```objc
typedef NS_ENUM(NSInteger, MDCTextControlLabelBehavior) {
  /** Indicates that the text control's label floats when text is entered. */
  MDCTextControlLabelBehaviorFloating,
  /** Indicates that the text control's label disappears when text is entered. */
  MDCTextControlLabelBehaviorDisappearing,
};
```
I kinda like the second one...

Reply:
Perhaps `MDCControlLabelBehavior`.

```objc
typedef NS_ENUM(NSInteger, MDCTextControlLabelBehavior) {
  /** Indicates that the control's label animates to a position above the control when text is entered. */
  MDCControlLabelBehaviorFloats,
  /** Indicates that the control's label disappears when text is entered. */
  MDCControlLabelBehaviorDisappears,
};
```

Reply:
Ehh though as I say that I think we should really try to keep these types from being publicly inherited across all of the components. Making this enum type too generic risks too much public coupling.

In that regard I think `MDCInputTextFieldLabelBehavior` is still preferable.

```objc
typedef NS_ENUM(NSInteger, MDCInputTextFieldLabelBehavior) {
  /** Indicates that the text field label animates to a position above the control when text is entered. */
  MDCInputTextFieldLabelBehaviorFloats,
  /** Indicates that the text field label disappears when text is entered. */
  MDCInputTextFieldLabelBehaviorDisappears,
};
```

Reply:
Ok, I don't mind if each of the contained input views in disguise have their own "label behavior" type as long as all of them can be used to determine the value for the internal shared "label state" type that's used in layout calculations and style application.

Reply:
Done in b151a4e.